### PR TITLE
Allow some non-integer built-in constants in gdscript

### PR DESCRIPTION
**Pull request by ghost**
I'm always missing named compile-time color constants for Colors in the editor. This adds them:

demo

Since adding full fledged static constants to gdscript (I started) might entail some longer discussions due to various side effects (time of initialization, change of byte code for singletons, etc.), this PR's implementaiton is just adding some syntactic sugar by expanding Color.xyz into Color(r, g, b, a) in the gdscript parser.

### CODE REVIEW

**Comment by akien-mga** [comment]
That feature is already implemented as `ColorN("red")`, I don't think we should add another way to do the same thing: http://docs.godotengine.org/en/latest/classes/class_@gdscript.html#class-gdscript-colorn

**Comment by poke1024** [comment]
`ColorN` doesn't have code completion.

**Comment by Calinou** [comment]
Considering that you might not want to use "pure" color values (which look unpleasant to the human eye), is this worth implementing? You could define constants using the `const` keyword in a script instead.

**Comment by leoddd** [comment]
Is there a way to add custom entries to ColorN in-game?
If not, then I'd say that this PR could be used instead of ColorN, just because it's actually more intuitive to access different built-in colors from the Color class, rather than a built-in function; and the code-completion is helpful as well, because I would've looked for "lime" instead of "yellowgreen".
If adding custom entries to ColorN is a thing that could be considered then I'd rather go that route though. Would make consistent color scheming across the game easier, at least.

**Comment by poke1024** [comment]
@leoddd That's my stance, basically. If there's a static, curated list of colors Godot ships with, I'd like to have it as editor hint in some way.

**Comment by bojidar-bg** [comment]
What about adding code completion to ColorN instead?
Also, I doubt such constants' resolution should be put in GDScript, it would be better if they are accessed similar to the numeric constants (to allow for, e.g. Vector3.up and the list as well).

**Comment by poke1024** [comment]
How would you add code completion to `ColorN`? It takes a string parameter - you'd need to add a completely different kind of code completion not available yet.

I agree `Vector3.up` (and others) would be nice. But that does suggest adding full static non-integer constants to gdscript, which I originally started but then stopped as it's a change I'm probably not fit to do in a way that will match Godot's compiler design; it probably involves changing Godot's byte code as it needs access to some sort of globals/singleton management, the only global being `self` AFAIK. One relatively simple approach would be having `static x = 3` internally rewrite into `static x = func return 3` and then internally rewrite `abc.x` into `abc.x()`; but that would not be a constant. It might be ok if we use this only internally and do not open definition of static constants to gdscript users.

On the other hand, if we do not expose this feature (definition on non-numeric static constants), one could generalize the approach of this PR to add `Vector.up` and similar builtin constants - I actually think that this approach wouldn't be so bad, as it's a small change and would allow for a set of non-integer static constants for some built-in classes.

**Comment by poke1024** [comment]
PR is now cleaned up and much generalized (changed PR title to reflect this).

To sum up this new PR: instead of making all static built-in constants integers (as in current master), built-in constants may now be other types (basically any `Variant` types that do not bloat the constant table, see below).

As an example, this implements `Vector3.UP`, `Vector3.DOWN`, and so on (besides `Color` names).

All non-color constants show in the docs just as integer constants do:

docs

Internally, in `GDScriptParser`, all constants are represented by a `ConstantNode` which AFAIK embeds the `Variant` in a constant table in the byte code - if this PR is followed, then there should be a guard to not allow any `Variant` but only some basic types like `Vector2`, `Vector3`, `String` that are ok to embed in this way.

What this is not: a full-fledged static constants solution for having any expression or object type as a static constant or that lets users define their own static constants.

File: core/variant.h [hunk 1]
```diff
@@ -395,9 +395,9 @@ class Variant {
 
 	void static_assign(const Variant &p_variant);
 	static void get_constructor_list(Variant::Type p_type, List<MethodInfo> *p_list);
-	static void get_numeric_constants_for_type(Variant::Type p_type, List<StringName> *p_constants);
-	static bool has_numeric_constant(Variant::Type p_type, const StringName &p_value);
-	static int get_numeric_constant_value(Variant::Type p_type, const StringName &p_value, bool *r_valid = NULL);
+    static void get_constants_for_type(Variant::Type p_type, List<StringName> *p_constants);
```
Comment (bojidar-bg) [inline core/variant.h]: Wrong indentation here, BTW.

**Comment by homer666** [comment]
This is just nitpicking, but shouldn't Vector3.BACK and Vector3.FORWARD be swapped around? I got the impression that in a right-handed coordinate system, from a perspective where x-positive would be "right" and y-positive "up", z-negative would be "forward". Here's a [reference](https://www.evl.uic.edu/ralph/508S98/coordinates.html) that I guess supports this.

**Comment by homer666** [comment]
Your `Vector2` constants also seem to have this problem; in the 2D view, `Vector2(0, -1)` is "up" and `Vector2(0, 1)` is "down".

**Comment by akien-mga** [comment]
> This is just nitpicking, but shouldn't Vector3.BACK and Vector3.FORWARD be swapped around? [...] Your Vector2 constants also seem to have this problem; in the 2D view, Vector2(0, -1) is "up" and Vector2(0, 1) is "down".

Proved the point I made every time such constants were asked for that they're not improving code quality. If people don't understand how the coordinate system works, having such helpers just makes code less readable and more error prone; whatever is Vector3.RIGHT when you're looking towards (5.24, -12.13, -2.0)?

Still, if people really want those, why not, they've been requested enough times to show that some people think differently from me ;) But I wouldn't extend GDScript features like non-integer built-in constants just for those.

**Comment by poke1024** [comment]
@homer666 Not an expert on this one, but here's some thoughts on this confusing topic.

If you'd pick a right-handed coordinate system and you'd define `forward` in a negative numeric direction, then `forward` is not a good name for decreasing in `z` IMHO.

The reason is, I believe, that most people tend to think in left-handed systems; and even though OpenGL is right-handed in object space, you often find camera views in OpenGL that are configured to left-handed systems (https://stackoverflow.com/questions/4124041/is-opengl-coordinate-system-left-handed-or-right-handed).

I agree with @akien-mga: all these terms lose their meaning if you're not in a standard camera frame. Only if you're in a standard camera in OpenGL object space, `forward` is `-z`.

I think you want`right`, `up` and `forward` to correspond to numerically increasing values in the respective axes, because that is what most people intuitively think. And that's left-handed.

Unity also follows this convention:
https://docs.unity3d.com/ScriptReference/Vector2.html
https://docs.unity3d.com/ScriptReference/Vector3.html

**Comment by poke1024** [comment]
Added a few more constants for illustration of what might make sense.

**Comment by homer666** [comment]
@poke1024 I do agree that left-handed would come naturally to most people, but I don't know if deviating from the way Godot works is really a good idea.

From what I can tell, Unity follows this convention because it actually uses a left-handed coordinate system for everything, so having `forward` be `+z` makes perfect sense there. Its 2D view also seems to just be the 3D world with the z-axis flattened, so `+y` is still `up` as it was in 3D.

Godot's two different coordinate spaces don't adhere to these two things. Identifying and understanding these little coordinate quirks have always been a requirement to using Godot beyond just plopping down Control nodes, so this won't really change anything beyond having to remember that some of the Vector constants are backwards...

**Comment by poke1024** [comment]
I understand, and indeed Unity seems to be consistenty left-handed. So, `Vector2.UP` should be what Godot uses as "up" in the standard 2d view, i.e. `(0, -1, 0)`. But then I also really get @akien-mga's point now.

I guess I should change the constants to the Godot way and users could be informed that these constants only make sense if they consistently set up their local transformation frames in a way that adheres to these convention.

The main question remains though if this PR will ever see the light of day :-)

**Comment by homer666** [comment]
I hope it does! Wrapping your head around the problem is a bump that people run into even without the constants; when I started playing around with 3D, my characters were facing `z+` which caused them to move "left" when `x+` increased, which tripped me up a little. I think that having this explicitly stated as a constant or something would help more than it would hinder honestly.

**Comment by poke1024** [comment]
Apart from experimenting with a few other constants (more like a brainstorm for anyone evaluating this PR) I've turned the color constants to upper case for consistency with other constants and made them appear in the docs:

color-docs

**Comment by akien-mga** [comment]
Thanks a lot for your contribution!
We are now entering a strict release freeze for Godot 3.0 and will only consider major bug fixes. We won't merge new features and enhancements anymore until 3.0 is released.

Moving this PR to the 3.1 milestone, to be reviewed once the release freeze is lifted. It could eventually be cherry-picked for a future 3.0.1 maintenance release if it doesn't change the user-facing APIs and doesn't compromise the engine's stability.

File: core/variant_call.cpp [hunk 2]
```diff
@@ -1811,6 +1860,54 @@ void register_variant_methods() {
...
 	_VariantCall::add_constant(Variant::VECTOR3, "AXIS_Z", Vector3::AXIS_Z);
+
+	_VariantCall::add_variant_constant(Variant::VECTOR3, "ZERO", Vector3(0, 0, 0));
+	_VariantCall::add_variant_constant(Variant::VECTOR3, "INF", Vector3(Math_INF, Math_INF, Math_INF));
+	_VariantCall::add_variant_constant(Variant::VECTOR3, "LEFT", Vector3(-1, 0, 0));
+	_VariantCall::add_variant_constant(Variant::VECTOR3, "RIGHT", Vector3(1, 0, 0));
+	_VariantCall::add_variant_constant(Variant::VECTOR3, "UP", Vector3(0, 1, 0));
+	_VariantCall::add_variant_constant(Variant::VECTOR3, "DOWN", Vector3(0, -1, 0));
+	_VariantCall::add_variant_constant(Variant::VECTOR3, "FORWARD", Vector3(0, 0, -1));
+	_VariantCall::add_variant_constant(Variant::VECTOR3, "BACK", Vector3(0, 0, 1));
```
Comment (aaronfranke) [inline core/variant_call.cpp]: Great work, although I'd also like to suggest adding `"ONE"` and maybe `"NEGONE"` for feature parity with the C# built-ins I added: https://github.com/godotengine/godot/blob/master/modules/mono/glue/cs_files/Vector3.cs#L276

**Comment by akien-mga** [comment]
This needs a rebase after #19264, but we should likely decide whether we want the feature or not. WDYT @vnen @reduz?

File: core/color.h [hunk 3]
```diff
@@ -187,6 +188,7 @@ struct Color {
 	static Color hex(uint32_t p_hex);
 	static Color html(const String &p_color);
 	static bool html_is_valid(const String &p_color);
+	static const Map<String, Color> &names();
```
Comment (reduz) [inline core/color.h]: I would simply not add this here, its not going to be used anyway, just put them together with the other constants inside Variant

**Review by reduz — CHANGES_REQUESTED** [review]

**Comment by reduz** [comment]
This is very cool, if you can change what mentioned in the review, we can merge it

File: modules/gdscript/gdscript_parser.cpp [hunk 4]
```diff
@@ -642,9 +642,8 @@ GDScriptParser::Node *GDScriptParser::_parse_expression(Node *p_parent, bool p_s
 			} else {
 
 				ConstantNode *cn = alloc_node<ConstantNode>();
-				cn->value = Variant::get_numeric_constant_value(bi_type, identifier);
-				cn->datatype = _type_from_variant(cn->value);
-				expr = cn;
+				cn->value = Variant::get_constant_value(bi_type, identifier);
+				cn->datatype = _type_from_variant(cn->value);expr = cn;
```
Comment (vnen) [inline modules/gdscript/gdscript_parser.cpp]: Slight nitpick: there shouldn't be 2 statements in the same line.

**Comment by karroffel** [comment]
Bump, there is a very small formatting issue, but I would love to see this merged! :blue_heart:

**Comment by karroffel** [comment]
Bump, there is a very small formatting issue, but I would love to see this merged! :blue_heart:

File: core/color.h (hunk 3 above)
Comment (poke1024) [reply core/color.h]: Done.

**Comment by akien-mga** [comment]
I agree with the feature itself, but I think the list of constants that was added is way overkill, this should be reviewed before 3.1 is out. Things like `BLACK_CIRCLE` seem hardly useful, quite arbitrary (I don't think we want to add constants for all of Unicode), and risky (they break generating RST documentation from XML currently).

**Comment by aaronfranke** [comment]
Why aren't the Color constants UPPERCASE?

Also, in C# they are in a `Colors` static class, so that they are separate from the `Color` struct's other static members (like the `Color8` constructor). While there are no other static members in GDScript's `Color` to warrant this change, it's worth considering `Colors.RED` etc for consistency.

**Comment by bojidar-bg** [comment]
(Note that changing the case of the constants might warrant a change in ColorN's documentation, due to #35581)

**Comment by ShivamMukherjee** [comment]
I can do performs those changes, I'm already invested in this.
https://github.com/godotengine/godot/pull/41019

**Pull request merged**